subsys/net/lib/http/http_server.c

```diff
@@ -0,0 +1,1038 @@
+			if (sendall(ctx_client->client_fd, not_found_response,
+				    strlen(not_found_response)) < 0) {
+				close_client_connection(ctx_server, client_index);
+			}
+		}
+		close_client_connection(ctx_server, client_index);
+		memset(ctx_client->buffer, 0, sizeof(ctx_client->buffer));
+		ctx_client->offset = 0;
+	}
+
+	return 0;
+}
+
+int handle_http_done(struct http_server_ctx *ctx_server, struct http_client_ctx *ctx_client,
+		      int client_index)
+{
+	LOG_DBG("HTTP_DONE_STATE");
+
+	close_client_connection(ctx_server, client_index);
+
+	return -errno;
+}
+
+int handle_http_frame_headers(struct http_client_ctx *ctx_client)
+{
+	LOG_DBG("HTTP_FRAME_HEADERS");
+
+	unsigned char response_headers_frame[16];
+
+	struct http_frame *frame = &ctx_client->current_frame;
+
+	print_http_frames(frame);
+
+	const char *method;
+	const char *path;
+
+	if (!has_upgrade_header) {
+		enum http_method method_h2c = parser.method;
+
+		method = http_method_str(method_h2c);
+		path = url_buffer;
+	} else {
+		method = header_parsing(ctx_client, HTTP_HPACK_METHOD);
+		path = header_parsing(ctx_client, HTTP_HPACK_PATH);
+	}
+
+	if (strcmp(method, "GET") == 0 && strcmp(path, "/") == 0) {
+
+		generate_response_headers_frame(response_headers_frame, frame->stream_identifier,
+						HTTP_HPACK_STATUS_200);
+
+		size_t ret = sendall(ctx_client->client_fd, response_headers_frame,
+				     sizeof(response_headers_frame));
+		if (ret < 0) {
+			LOG_ERR("ERROR writing to socket");
+			return -errno;
+		}
+
+		size_t content_size = sizeof(content_200);
+
+		send_data(ctx_client->client_fd, content_200, content_size, HTTP_DATA_FRAME,
+			  HTTP_FLAG_END_STREAM, frame->stream_identifier);
+	} else {
+		generate_response_headers_frame(response_headers_frame, frame->stream_identifier,
+						HTTP_HPACK_STATUS_404);
+
+		size_t ret = sendall(ctx_client->client_fd, response_headers_frame,
+				     sizeof(response_headers_frame));
+		if (ret < 0) {
+			LOG_ERR("ERROR writing to socket");
+			return -errno;
+		}
+
+		size_t content_size = sizeof(content_404);
+
+		send_data(ctx_client->client_fd, content_404, content_size, HTTP_DATA_FRAME,
+			  HTTP_FLAG_END_STREAM, frame->stream_identifier);
+	}
+
+	ctx_client->server_state = HTTP_FRAME_HEADER_STATE;
+
+	return 0;
+}
+
+int handle_http_frame_priority(struct http_client_ctx *ctx_client)
+{
+	LOG_DBG("HTTP_FRAME_PRIORITY_STATE");
+
+	struct http_frame *frame = &ctx_client->current_frame;
+
+	print_http_frames(frame);
+
+	ctx_client->server_state = HTTP_FRAME_HEADER_STATE;
+
+	return 0;
+}
+
+int handle_http_frame_continuation(struct http_client_ctx *ctx_client)
+{
+	LOG_DBG("HTTP_FRAME_CONTINUATION_STATE");
+	ctx_client->server_state = HTTP_FRAME_HEADERS_STATE;
+
+	return 0;
+}
+
+int handle_http_frame_settings(struct http_client_ctx *ctx_client)
+{
+	LOG_DBG("HTTP_FRAME_SETTINGS");
+
+	struct http_frame *frame = &ctx_client->current_frame;
+
+	print_http_frames(frame);
+
+	if (!settings_ack_flag(frame->flags)) {
+		ssize_t ret =
```

Comment:
My previous comment on this seems to have been lost in action (https://github.com/zephyrproject-rtos/zephyr/pull/59669#discussion_r1275834891),  but again, we should not be sending a SETTINGS frame as a response to the client's settings frame. To comply with the RFC, it should be sent as soon as connection is established (so when the preface indicates we deal with HTTP2, we should send it rightaway then).

Reply:
I tried to do this, but the client was blocked. I'll look into it further.